### ISSUE TITLE
Send a PKI from the client to the server

### ISSUE THREAD

**Issue by justin-bits** [cypress-io/cypress#2694]
Need a way to send a PKI from the client to the server when it is requested by the server on login (two way SSL).

On connecting to our application, the server requests a PKI / personal certificate from the client.  This PKI is unique to the user; it is installed in the browser so that they can authenticate to websites with it.

The browser prompts the user to “Select a certificate to authenticate yourself”, where the user can select OK to send it.

This prompt/dialog isn’t appearing in Test Runner.  Instead, our custom error message indicating that the user did not send a PKI is displayed.

I attempted to use a group policy to force all instances of Chrome to auto select a certificate, but the PKI still isn’t sent from the client to the server. The group policy does work for all other instances of Chrome that aren’t created by Cypress.

**Comment by jsalankey** [comment]
This is a critical piece of functionality for our ability to use Cypress as our testing framework.

**Comment by sforner405** [comment]
We also need to add a specific certificate to communicate with the server.
This feature is necessary if we want to use cypress.

**Comment by brian-mann** [comment]
I've reviewed the PR here #2740 and understand the implications of this and have some feedback.

I see three areas that need to be addressed:

1. Would attaching the client certs only on `cy.visit()` actually fix the issue? Wouldn't the browser need to send the certs on **all** requests, not just the first?

2. We'd also need to ensure this works on `cy.request()` too since that is made outside of the browser.

3. I need to review how browser normally accept the PKI certificates as command line arguments or through group policy to understand the various ways in which they can be configured. These would at the least need to be configurable at the `cypress.json` level, and likely need some way to control which domain(s) they are sent to - enabling you to have multiple certificates for multiple domains.

Although we could do everything at the network layer transparently - I believe it likely makes more sense to utilize the browser agent itself to handle automatically attaching these instead of at the proxy layer. That way problems would bubble up to the dev tools in the browser and would be more immediately apparent.

**Comment by sforner405** [comment]
Even in a CI environment it is more difficult to add this to a browser instance. 
Had a lot of issues there recently.

**Comment by matt-rhys-jones** [comment]
Thanks for reviewing :) Just to provide a bit of background in regards to your first point, our reasoning for focusing on `cy.visit()` was because it was impossible for us to perform *any* form of e2e testing on the page under test without some way of utilising certificates.

By allowing certs to be passed through via `cy.visit()` we could at least load the initial page, then use fixtures for subsequent requests the page made. While this doesn't give us a true e2e picture, it would allow us to have fairly comprehensive front-end tests and significantly enhances our ability to adopt Cypress while further support for PKI was being considered/worked on. We saw it very much as a potential first step, but one that would have immediate and significant benefit to us at the BBC.

In regards to your third point, my understanding of how PKI typically works is that a certificate is unique to a single user/machine, but can be used to authenticate across multiple domains that a corporation manages. With this in mind, certainly for us, the ability to configure just a single certificate in `cypress.json`, or by another means, would be enough for us to make full use of what Cypress has to offer.

We'd more than happily accept whatever implementation you felt was best, but hopefully this sheds some more light on what we're trying to achieve.

**Comment by craigkj** [comment]
@brian-mann Hi there. I worked with @matt-rhys-jones on this and we would like to keep this PR moving. Did you come to a conclusion on what you thought the best approach was? Thanks for the feedback btw. Cheers.

**Comment by brian-mann** [comment]
This PR would need to be updated to reflect / address the issue in a more comprehensive way. I would prefer coming up with a solution that:

1. works across all browsers
2. works for `cy.visit`, `cy.request`, and all other network traffic the browser makes

Updating just `cy.visit()` would not really address this issue - and my preference would be either adding command line arguments or `cypress.json` configuration to address this.

@matt-rhys-jones 	and @craigkj 	what would be most helpful is if you can find us examples of using other testing tools to accommodate this need - selenium/webdriver, puppeteer, etc. 

When I looked I did not see a way to launch chrome with arguments that would add these certificates so I cannot find a comparable way in which this could be achieved. If there are existing solutions we can piggy-back off of their ideas.

To be clear, without an easy way to configure each browser means that we'd be looking at implementing this at the network layer behind the scenes in a way that the browser doesn't know that it's happening. That'll work across all browsers but may have unintended effects, since I'm not super experienced with PKI in general, although I understand how it works.

Additionally it seems that whatever way we come up with would have to be targeted on a domain basis, and possibly also expose specific sets of configuration.

**Comment by matt-rhys-jones** [comment]
Hi @brian-mann - thanks for your thoughts. I agree that updating `cy.visit()` only partially addresses the issue (and was only intended to do so initially).

In regards to existing solutions, I haven’t come across many in the past, and certainly not any *elegant* ones. Most seem to involve automating dialogue box clicks with device automation tools. Selenium does allow the web driver to be configured to custom Firefox profiles with the various certificates provided but as you say this isn't possible across all browsers.

These solutions are inherently flakey, you'd know more than me but personally I don’t think it would be easy to come up with a *reliable* cross browser solution that automated the various dialogue boxes (or configs for headless browsers) to support PKI. This would also require continued maintenance as new versions of browsers emerge, and documentation updates too.

I think that for this to work reliably it would need to be done at the network level.

My understanding is that SSL operates at the Session/Presentation layer of the OSI model rather than Application layer. While the browser *can* be configured to support PKI in it's entirety, I believe the Operating System would *typically* be responsible for housing and trusting Root Certificates / Certificate Authorities, and the browser would verify against these. Either way, how the browser ultimately handles the HTTPS response should remain unchanged as that’s at the Application layer.

This is probably an over-simplistic viewpoint and I"m sure the separation between OSI layers in the browser isn't as clear-cut as that, but that's my understanding of how it *should* work.

Therefore I think implementing this transparently at the network layer of Cypress makes more sense as the browser shouldn't care whether the original request was being made to a PKI domain or not providing it gets a valid HTTPS response from it's original request.

If we were worried about side effects it might be sensible to implement PKI support behind an opt-in feature flag or switch (e.g. cypress run —enablePKI).  If the flag is enabled then we could check the configuration for certificate/domain combinations and pass the certificates on with the request to the relevant domains transparently.

 I’d imagine the most challenging part would be maintaining the standard of Cypress’s error handling and messages. Errors returned from `openssl` are not particularly friendly and I’d imagine we’d need to wrap around these so that we can provide clear guidance to users on which part of the PKI authentication went wrong (e.g. badly formatted certs, missing CA, missing certificate/key, unable to validate certificate against CA). I’m not sure how easy that would be, and it would require more understanding of the TLS spec and `openssl` internals/errors which again would take time.

If you think the above approach is viable, and Cypress would be happy to provide support with PR reviews and some guidance (if the docs fail me) then I would be happy to have a go at implementing this in my spare time.

Alternatively, if you wanted to add PKI support to your roadmap and develop in house then I would be happy to test it out.

Either way, it's definitely a feature we feel would be valuable for us at the BBC, add value to Cypress as a product, and address the original issue that was raised.

Please do let me know how you'd like to proceed! :)

**Comment by sforner405** [comment]
Seems that our problem is also not fixed by the PR.
We need to add the certificate to the XHR requests made by the browser.
So the requirement 
"works for cy.visit, cy.request, and all other network traffic the browser makes"
is important for us.
Best thing would be to have something like:
`{
  "certificates": {
    "baseurl": {
      "key": "keypath",
      "cert": "certpath"
    }
  }
}`

**Comment by matt-rhys-jones** [comment]
Yes, it should definitely be implemented for all requests. However you can at least use [fixtures](https://docs.cypress.io/guides/guides/network-requests.html#Fixtures) for your subsequent XHR requests as a middle ground until we have a more rounded solution.

**Comment by sforner405** [comment]
I can see it was added to Sprint 11, but never got fixed or added to the subsequent sprint milestones.
Is this still something you are looking into?

**Comment by jennifer-shehane** [comment]
The scope of this issue has been expanded - so work on this has been postponed until the implementation details are decided first. See https://github.com/cypress-io/cypress/issues/2694#issuecomment-439970547

**Comment by clevrtec** [comment]
Realize this is closed, just wanted to drop for future reference how we did it.  I'm a government software engineer and ran into this problem today, decided to just leave Cypress as-is and use a implement a reverse proxy instead.  Basically expose different proxy ports with different scenarios: Valid, invalid pass, revoked, tampered, etc.  That way Cypress (or anything else can just get the scenarios it needs via a port.

We generate the SSL certs dynamically and inject them into the test environment.  Also note, this app uses web sockets so I had to add a handler for that, it would be even simpler without.

```javascript
const httpProxy = require('http-proxy');
const fs = require('fs');
const http = require('http');
const morgan = require('morgan')('dev');

proxyBuilder('Valid cert', 'client-valid.p12', 'test', 25000);
proxyBuilder('Invalid cert password', 'client-valid.p12', 'bad-password', 25001);
proxyBuilder('Untrusted cert', 'client-untrusted.p12', 'test', 25002);
proxyBuilder('Revoked cert', 'client-revoked.p12', 'test', 25003);
proxyBuilder('Expired cert', 'client-expired.p12', 'test', 25004);
proxyBuilder('Tampered cert', 'client-tampered.p12', 'test', 25005);

function proxyBuilder(testTitle, certPath, certPass, targetPort) {

  try {

    const proxy = new httpProxy.createProxyServer({
      target: {
        protocol: 'https:',
        host: 'localhost',
        port: process.env.SERVER_PORT,
        pfx: fs.readFileSync(`${__dirname}/../ssl/${certPath}`),
        passphrase: certPass,
      },
      secure: false,
      changeOrigin: true,
    });

    const proxyServer = http.createServer((req, res) => {
      console.log(testTitle);
      morgan(req, res, () => null);
      proxy.web(req, res);
    });

    proxyServer.on('upgrade', (req, socket, head) => {
      console.log(testTitle);
      morgan(req, socket, () => null);
      proxy.ws(req, socket, head);
    });

    proxyServer.listen(targetPort);

  } catch (e) {
    console.error(`Problem starting ${testTitle}`);
    console.error(e);
  }

}
```

**Comment by jennifer-shehane** [comment]
@clevrtec This issue is not closed. It is part of our roadmap, but has not had any work done on it at this time.

**Comment by c32hedge** [comment]
In my use case, I would like to use cy.request to make API calls that require certificate-based authentication. Being able to use certificates at all would be a minimum need for me to be able to test our UI when we make changes via the API. However, I'd like to be able to use Cypress to test the API itself, and to do that I would want to test using different certificates, so just setting one in a config file wouldn't be ideal.

**Comment by Brett-Holmes** [comment]
We also need this feature to be able to use Cypress at all. Any Updates? Thank you!

**Comment by flotwig** [comment]
@Brett-Holmes Not currently being worked on, though we'd be happy to accept a PR that implements this functionality. The workaround that @clevrtec is the way to do this in the meantime.

**Comment by kmaverick** [comment]
Did you guys get an answer to the original question?

**Comment by flotwig** [comment]
I did some experiments today to try to understand what a thorough solution to this would look like: https://gist.github.com/flotwig/12b2f070fd627f48fa7681802e6c36de

I noticed that we can't just send a bunch of PKI certs and hope that Node.js uses the correct one - we have to use the correct cert for each server: https://gist.github.com/flotwig/12b2f070fd627f48fa7681802e6c36de#file-pki-js-L84-L94

I can see why a user would want to configure this in `cypress.json`, if they aren't particularly interested in testing how PKI affects their website, they just want to be able to load it. That could look something like this:

```js
{
	/** rest of the cypress.json... **/
	"clientCertificates": [
		{
			"url": "http://megacorp.net/secure/*",
			"cert": "./cypress/support/my-cert.crt",
			"key": "./cypress/support/my-private-key.key"
		},
		{
			// this is an example of a chained client certificate, which is apparently possible
			// see `cert` and `key` on https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options 
			"url": "http://megacorp.net/chained-secure/*",
			"cert": [
				"./cypress/support/my-first-cert.crt",
				"./cypress/support/my-second-cert.crt"
			],
			"key": [
				"./cypress/support/my-first-private-key.key",
				"./cypress/support/my-second-private-key.key",
			]	
		},
		// additional configurations can go here, the first matching `url`'s `cert` 
		// and `key` will be used for any visits that match that `url`
	]
}
```

I think we should also expose an option on `cy.visit` and `cy.request` that follows the same syntax, so users can test what happens when they use invalid client certificates on their website:

```
cy.visit({
	url: 'http://megacorp.net/secure/',
	cert: "./cypress/support/my-cert.crt",
	key: "./cypress/support/my-private-key.key"
})
```

Same sort of syntax for `cy.request`. This would override the `clientCertificates` from `cypress.json`, but only for that exact visit or request, not for any subresources.

Alternatively, using a `cert` and `key` with `cy.visit` could cause all requests that start with the same origin to use that client certificate for the duration of the current test, which would allow users to test subresources getting PKI as well.

---

I've never had to work with an implementation of PKI in the wild, so I would appreciate any comments regarding this design, particularly concerning if sending a PKI with `cy.visit` should apply to all requests in the same origin or not.

**Comment by blankey1337** [comment]
@flotwig this looks great, what do we need to get a PR for this?

**Comment by matt-rhys-jones** [comment]
Apologies I haven't had time to keep up to date with this issue. @flotwig the PR looks great :)

I think we'd also need `ca` to be added to the list as many large organisations have their own certificate authority which the certs need to be validated against (as opposed to a Root CA which the underlying OS would already know about).

**Comment by blankey1337** [comment]
Would love to use this on my team, but sadly this is a blocker for us.

**Comment by justin-bits** [comment]
I ended up creating a script to start a proxy server to workaround this limitation. Check out https://github.com/http-party/node-http-proxy

**Comment by pablogiroud** [comment]
@jennifer-shehane Good morning! i'm working on API tests and have the same issue with certificates. Can you confirm if you will work in some solution for this? Like @jsalankey mentioned, this is critical to us.
I have two months working with cypress and love it.

> I did some experiments today to try to understand what a thorough solution to this would look like: https://gist.github.com/flotwig/12b2f070fd627f48fa7681802e6c36de
> 
> I noticed that we can't just send a bunch of PKI certs and hope that Node.js uses the correct one - we have to use the correct cert for each server: https://gist.github.com/flotwig/12b2f070fd627f48fa7681802e6c36de#file-pki-js-L84-L94
> 
> I can see why a user would want to configure this in `cypress.json`, if they aren't particularly interested in testing how PKI affects their website, they just want to be able to load it. That could look something like this:
> 
> ```js
> {
> 	/** rest of the cypress.json... **/
> 	"clientCertificates": [
> 		{
> 			"url": "http://megacorp.net/secure/*",
> 			"cert": "./cypress/support/my-cert.crt",
> 			"key": "./cypress/support/my-private-key.key",
> 			"passphrase": "some optional passphrase for the key",
> 		},
> 		{
> 			// this is an example of a chained client certificate, which is apparently possible
> 			// see `cert` and `key` on https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options 
> 			"url": "http://megacorp.net/chained-secure/*",
> 			"cert": [
> 				"./cypress/support/my-first-cert.crt",
> 				"./cypress/support/my-second-cert.crt"
> 			],
> 			"key": [
> 				"./cypress/support/my-first-private-key.key",
> 				"./cypress/support/my-second-private-key.key",
> 			]	
> 		},
> 		{
> 			// an example of using a pfx/pkcs12 key and cert chain combined
> 			"url": "http://megacorp.net/pfx-secure/*",
> 			"pfx": "./cypress/support/my-pfx.pfx",
> 			"passphrase": "some optional passphrase for the key",
> 		},
> 		// additional configurations can go here, the first matching `url`'s `cert` 
> 		// and `key` will be used for any visits that match that `url`
> 	]
> }
> ```
> 
> I think we should also expose an option on `cy.visit` and `cy.request` that follows the same syntax, so users can test what happens when they use invalid client certificates on their website:
> 
> ```js
> cy.visit({
> 	url: 'http://megacorp.net/secure/',
> 	cert: "./cypress/support/my-cert.crt",
> 	key: "./cypress/support/my-private-key.key"
> })
> ```
> 
> Same sort of syntax for `cy.request`. This would override the `clientCertificates` from `cypress.json`, but only for that exact visit or request, not for any subresources.
> 
> Alternatively, using a `cert` and `key` with `cy.visit` could cause all requests that start with the same origin to use that client certificate for the duration of the current test, which would allow users to test subresources getting PKI as well. Could be a flag to switch to this behavior, something like `usePkiForOrigin: 'http://megacorp.net/secure/*'` to indicate what URLs should match
> 
> I've never had to work with an implementation of PKI in the wild, so I would appreciate any comments regarding this design, particularly concerning if sending a PKI with `cy.visit` should apply to all requests in the same origin or not.

**Comment by jdmairs** [comment]
I have no experience with e2e testing but I've seen the twitter of folks I respect that say, "Write more integration tests then Unit Tests...and use Cypress.  I need pki certificate support as well.

**Comment by kmaverick** [comment]
What pki help do you need? I can help

**Comment by navjeetc** [comment]
I also have a need to support client side certificates.It will be great to Cypress support for client certs.  @justin-bits can you provide some more info on how did you use proxy server to implement a work around? My web apps are written in Ruby on Rails or have Vue.js(Nuxt.js) front end calling APIs written in Rails. We use Nginx as frontend web server with Passenger as Ruby application server.

**Comment by kmaverick** [comment]
I used apache to handle proxy and certs. If you are using nginx check the
configuration file. You will need to research what ssl mods nginx has. It
routes the certs to your application by defined url.

**Comment by navjeetc** [comment]
@kmaverick Nginx is configured to forward certs and it works in the web browser (for now printing cert CN in php script and it  prints correct value). But when cypress runs the test, CN is blank.

**Comment by GCHQDeveloper911** [comment]
Forgive me here folks, but am I missing something? Is there a solution to this now available? We're really keen to use Cypress, but are seeing SSL errors as seen in #4394 - apparently because our client cert isn't being passed in on test launch. As for many others, this basically blows Cypress out of the water as a non-starter.... Does cy.visit now provide a mechanism to specify the cert  and key(that I've missed the docs for?)... Or is there another way of passing them in at test launch (eg, specify in cypress.json?). FWIW, we're currently on Cypress 4.10.0, with Node 12.18.2. 

TIA!

**Comment by kmaverick** [comment]
Are you trying to connect to a remote server or web service or something?

**Comment by vijar50** [comment]
Hello I am running into the same issue. We are using PKI and need a way to handle Certs in Cypress. Is there any work around available at the moment?

**Comment by onetrickwolf** [comment]
@vijar50 and @GCHQDeveloper911 the solution of creating a proxy to pass the client cert works fine for us for now. This solution shown earlier in the thread works just fine for us: https://github.com/cypress-io/cypress/issues/2694#issuecomment-464270224

It would be nice though if Cypress could support this though as adding a proxy is a bit clunky and confusing to some people.

**Comment by GCHQDeveloper911** [comment]
Thanks @onetrickwolf - So if I'm understanding correctly.... I use the code above to create a standalone node app, but changing target.host and target.port to point at the site I'm testing, and then in my cypress test I'll have something like cy.visit('https://localhost:25000') ??

I've tried the above (with a dummy self created p12 for a locally running test app), but even ignoring cypress, and trying to browse to https://localhost:25000, I get 'the site can't provide a secure connection (err_ssl_protocol_error)'?

What am I not understanding? :) Thanks again!

**Comment by onetrickwolf** [comment]
@GCHQDeveloper911 yes that is correct but I believe you would hit **http** NOT **https**. The idea is that the proxy is setting up a completely unsecured connection for Cypress to use that proxies to the secured connection.

**Comment by GCHQDeveloper911** [comment]
If I hit (even manually, without cypress) http://localhost:25000, my proxy app crashes out with:
```Error: write EPROTO 139744977135488:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:```

**Comment by GCHQDeveloper911** [comment]
Hi @onetrickwolf - just an update to say thanks again for this - it works... to a point - hopefully you can help get us beyond that point?

So if I set my cypress test to cy.visit('https://my.test.url') then it bombs with the SSL v3 handshake error.
If I set it to cy.visit('http://localhost:25000') it loads the page correctly. (obviously my proxy is configured with my cert etc

However, there are parts of our app that make REST calls to a back end system to obtain an auth token... and those calls are failing with 'no client certificate was presented'

Curiously, if I manually browse to localhost:25000, I don't experience the same problem - it is only when run via Cypress.

Thoughts?

**Comment by GCHQDeveloper911** [comment]
Where I work we now have a patched version of 6.0.1 that, together with a pair of custom commands and some appropriate certificate config entries in cypress.env.json / cypress.json gives us cy.visitWithCerts and cy.requestWithCerts, and removes the need for reverse proxies and so on.... I'm thinking it'd be good to eliminate the custom commands bit and add the 'WithCerts' commands to the cypress code and push up to github as a PR in the latest version of Cypress - but before I did that, is this something that would get support from the Cypress authorities? Would 2 new Cypress commands like this be dismissed as unwanted?

**Comment by DannyFeliz** [comment]
I am still waiting for this. I need this to run my test suite in the dev env.

**Comment by kmaverick** [comment]
What are you waiting for?

**Comment by CMalanaphyStatpro** [comment]
> Where I work we now have a patched version of 6.0.1 that, together with a pair of custom commands and some appropriate certificate config entries in cypress.env.json / cypress.json gives us cy.visitWithCerts and cy.requestWithCerts, and removes the need for reverse proxies and so on.... I'm thinking it'd be good to eliminate the custom commands bit and add the 'WithCerts' commands to the cypress code and push up to github as a PR in the latest version of Cypress - but before I did that, is this something that would get support from the Cypress authorities? Would 2 new Cypress commands like this be dismissed as unwanted?

Hi @GCHQDeveloper911, Did you get any feedback on your changes? if not could you send an example of the custom commands and certificate config entries in cypress.env.json

**Comment by GCHQDeveloper911** [comment]
Hi @CMalanaphyStatpro - Didn't really get any feedback, but I'm working on it anyway - give me a bit to sort a couple of issues!

**Comment by jennifer-shehane** [comment]
@GCHQDeveloper911 We would be open to a PR for this. It may help accelerate the work. We would still likely need to do some additional work on top of it. We're not sure exactly how the API would look, whether it would be those commands exactly. We would also want to have full cross browser support before an official release. Feel free to open a PR in 'draft' and let us know if you need any review.

A way to get this out quickly, without the Cypress team's complete buy in, would be to create a [plugin](https://on.cypress.io/plugins-guide) that people can use.

**Comment by GCHQDeveloper911** [comment]
Hi @jennifer-shehane Some colleagues of mine have picked this up, and I think are close to a "pr-worthy" solution. However, I'll ping them a link to this thread, to highlight the plugin suggestion. Thanks!

**Comment by GCHQDeveloper911** [comment]
Hi @jennifer-shehane , it's still something of a WIP, with docs, and we're still working on a PKI test env / demo environment, but if you want to take a look at how it's shaping up so far, before we submit a PR, please do click here: https://github.com/GCHQDeveloper911/cypress

**Comment by jonminter** [comment]
@GCHQDeveloper911 Wondering about the progress your colleagues are making with this. Have a need for cert based authentication with Cypress for the project I am working on. Currently using a reverse proxy but would be interested in a different approach that doesn't require that.

I'd be willing to help with the implementation. The plugin idea sounded promising.

**Comment by GCHQDeveloper911** [comment]
Hi @jonminter - I just spoke to the guys, they're just doing some work to re-baseline against 6.5.0... Hopefully get a PR in soon...

Cheers

**Comment by thirstycoda** [comment]
Hi @GCHQDeveloper911,

We were hoping to use Cypress with certificates to impersonate different users at the same URL to test differing functionality/permissions. It looks like your PR introduces the ability to use the same certificate for all tests that relate to a particular URL. Am I right in thinking Cypress still won't support our requirement after your PR or is there a way to achieve what we need?

Thanks

**Comment by jonminter** [comment]
@GCHQDeveloper911 @thirstycoda 

I have the same requirements and question after reading over the PR. I think this could be accomplished by adding a method to the Cypress API to set or override the configured client certificate for the request. Something like:

```
cy.clientPkiCerificate({
  "pfx": "/home/tester/certs/cert.pfx",
  "passphrase": "/home/tester/certs/pfx-passphrase.txt"
});
```

**Comment by GCHQDeveloper911** [comment]
Hi @jonminter & @thirstycoda ....

We completely agree with you this would be a very nice further feature. However, for now our focus was a design that did not alter the 'test-facing' APIs so that it could be added to Cypress without breaking Cypress users' existing tests - key to getting the Cypress authorities to accepting the change. 

The guys I've been working with that have done the hard work(!) on this have more changes coming / planned, including cert validation on start up. I'm sure they'll also look at your 'cert switching' suggestions too. But for now, the view is "keep it simple" and "get the basics approved".

Hope this helps :)

**Comment by flotwig** [comment]
+1 @GCHQDeveloper911, per-test PKI config can be added at a later date. See https://github.com/cypress-io/cypress/issues/2694#issuecomment-503267977 for some additional discussion on API design.

**Comment by GCHQDeveloper911** [comment]
@flotwig My colleagues tell me they are looking into options for this sort of thing. They hope to have some sort of update early next week.

**Comment by Nastaran200** [comment]
> @GCHQDeveloper911 Wondering about the progress your colleagues are making with this. Have a need for cert based authentication with Cypress for the project I am working on. Currently using a reverse proxy but would be interested in a different approach that doesn't require that.
> 
> I'd be willing to help with the implementation. The plugin idea sounded promising.


I need immediate help on getting the softcert to work with cypress automation.  new to the tool. Can you please provide details as to how you accomplish this using revers proxy?  Greatly appreciated!

**Comment by steverb1** [comment]
@jennifer-shehane Has this been addressed yet?  If not, do you have any timeframe for it?
This is something we badly need also, before our company can adopt Cypress for use.

**Comment by Nastaran200** [comment]
> pair of custom commands and some approp
would you pls provide mre details in the setup of this. I am currently looking in doing this with no luck.

**Comment by Nastaran200** [comment]
> > Where I work we now have a patched version of 6.0.1 that, together with a pair of custom commands and some appropriate certificate config entries in cypress.env.json / cypress.json gives us cy.visitWithCerts and cy.requestWithCerts, and removes the need for reverse proxies and so on.... I'm thinking it'd be good to eliminate the custom commands bit and add the 'WithCerts' commands to the cypress code and push up to github as a PR in the latest version of Cypress - but before I did that, is this something that would get support from the Cypress authorities? Would 2 new Cypress commands like this be dismissed as unwanted?
> 
> Hi @GCHQDeveloper911, Did you get any feedback on your changes? if not could you send an example of the custom commands and certificate config entries in cypress.env.json

Hello @CMalanaphyStatpro  same here.  If you could provide more details as to how you accomplished this greatly appreciated.

**Comment by markslater** [comment]
I'm really happy to see progress on this feature.  In common with other commenters, I am working at an organisation that uses client certificates extensively, so to date, our Cypress tests haven't been truly representative of real world usage.

I want to second @thirstycoda's [comment](https://github.com/cypress-io/cypress/issues/2694#issuecomment-786074772): being able to configure Cypress with a single client certificate is a big step forward, but really unlocking all client certificate based testing will require the capability to set certificates per call.  In addition to the example @thirstycoda gave, it would be good to test the rainy day cases, like users presenting an expired certificate, or an invalid certificate, or no certificate, etc.

It would be great if this could be given consideration soon after #15179 (I also wonder if #15179 does actually make switching between client certificates possible, by changing the file referred to in the config on the fly).

**Comment by steverb1** [comment]
While I agree more capability around certificates will be helpful, let's get this first increment to the finish line (released) before we worry about more.  I wouldn't want to see this delayed, by starting to look at more.

**Comment by pcj** [comment]
This is kind of a big deal.

Trying to upgrade our tests from cypress 4.2 to 7.3.0 and hitting this.  AFAIK, Cypress does not surface TLS client certificate errors.  I got so frustrated by this (I didn't know what the underlying issue was... why SSO authentication no longer worked) I started a sample project using playwright.  Sure enough, the first chromium test I ran popped up a dialog asking me what client certificate I'd like to use.

Here's the workaround for playwright: https://github.com/microsoft/playwright/issues/1799

**Comment by upsampled** [comment]
@jeff-mccoy thanks for sharing that work around. Do you mind clarifying how you are using it: 
* Are you running that code as a service? or part of some cypress mechanic?
* Is DNS resolved at the proxy or the cypress client? 
* Does the `host: 'localhost'` line mean that all requests to the proxy are redirected to that domain name? Does changing it to `my-dev.corpA.com` mean that I can tell cypress to send requests to `127.0.0.1:25000` and they will then be redirected to `my-dev.corpA.com:443`?


I see @GCHQDeveloper911 is the hero hammering out the PR with this. Thank you for doing that. 

### Current Use-case  
Just to generally add to the requirements/expectations, here is my setup:
* mostly isolated test network with its own root DNS server (so we can actually use Internet DNS names)
* only way to get data to/from this network is a single SOCKv5h proxy. The 'h' just means I force DNS resolution to occur at the proxy so that domain name are resolved on a real nameserver. `git`, `go`, `curl`, `wget` all handle it without issue.

The testbed allows for a real mtls connection. While this could be done with just having a local nameserver override a record, that is more complex and does not have the overall assurance of a single proxy enforcing internet access.

Right now the main issues with cypress integration are:
* mtls  (covered in this ticket)
* Use of a proxy prevents the logging of test results, also SOCKSv5 is not supported but I have no problem adding other proxies

I believe that the @jeff-mccoy 's solution can actually handle this, as long as I run `cypress` on the SOCKv5 machine (real internet setup) and the BASEURL given to cypress is just an IP:Port (hopefully the https proxy will use the isolated DNS server, else I will just have to play games with `/etc/hosts`).
